nengo/synapses.py

```diff
@@ -221,123 +197,123 @@ def evaluate(self, frequencies):
         y = np.polyval(self.num, w) / np.polyval(self.den, w)
         return y
 
-    def make_step(self, shape_in, shape_out, dt, rng, y0=None,
-                  dtype=np.float64, method='zoh'):
-        """Returns a `.Step` instance that implements the linear filter."""
+    def _get_ss(self, dt, method='zoh'):
+        A, B, C, D = tf2ss(self.num, self.den)
+
+        # discretize (if len(A) == 0, filter is stateless and already discrete)
+        if self.analog and len(A) > 0:
+            A, B, C, D, _ = cont2discrete((A, B, C, D), dt, method=method)
+
+        return A, B, C, D
+
+    def allocate(self, shape_in, shape_out, dt,
+                 y0=None, dtype=np.float64, method='zoh'):
         assert shape_in == shape_out
 
-        num, den = self.num, self.den
-        if self.analog:
-            num, den, _ = cont2discrete((num, den), dt, method=method)
-            num = num.flatten()
-
-        if den[0] != 1.:
-            raise ValidationError("First element of the denominator must be 1",
-                                  attr='den', obj=self)
-        num = num[1:] if num[0] == 0 else num
-        den = den[1:]  # drop first element (equal to 1)
-        num, den = num.astype(dtype), den.astype(dtype)
-
-        output = np.zeros(shape_out, dtype=dtype)
-        if len(num) == 1 and len(den) == 0:
-            return LinearFilter.NoDen(num, den, output)
-        elif len(num) == 1 and len(den) == 1:
-            return LinearFilter.Simple(num, den, output, y0=y0)
-        return LinearFilter.General(num, den, output, y0=y0)
-
-    @staticmethod
-    def _make_zero_step(shape_in, shape_out, dt, rng, y0=None,
-                        dtype=np.float64):
-        output = np.zeros(shape_out, dtype=dtype)
-        if y0 is not None:
-            output[:] = y0
-
-        return LinearFilter.NoDen(np.array([1.]), np.array([]), output)
+        A, B, C, D = self._get_ss(dt, method=method)
+        X = np.zeros((A.shape[0],) + shape_out, dtype=dtype)
+
+        if y0 is not None and len(X) > 0:
+            y0 = np.array(y0, copy=False, ndmin=1)
+            X[:] = np.linalg.solve(np.eye(len(A)) - A, B * y0)
```

Comment:
Should `y0` be renamed to `u0` to indicate that it reflects a steady-state input and not an output? 

That is, this is solving:

`x = Ax + Bu` in order to find the stable-point `x` that corresponds to a constant input of `u`? The docstring can also be updated to reflect this.

But more importantly this has no solution if the system is exponentially unstable, which can happen in perfectly valid cases. For example, the integrator `1/s` is commonly used as a synapse by members of the lab (and users of nengolib) as an elegant way of performing perfect integration across some connection (often to compute the ideal output of an integrator, or as a dynamical primitive for some larger system). In this case we have `A = 0` before discretization and `A = 1` after discretization, hence:

`x = x + Bu` 

which has a solution of `x = 0` if and only if `u = 0`, otherwise it has no solution. This makes sense because an integrator never settles to a constant state given a non-zero input.

More detailed discussion is here for reference: https://github.com/nengo/nengo/issues/1124

Reply:
It seems to me that #1124 is the real issue we have to address here. But if renaming it to `u0` makes it clearer to you, I'm fine with that. It will be an API change, but I think not many people are using that parameter.

Reply:
In #1535, `y0` now actually represents the steady-state output. We still have the issue of it not working for some synapses, but that's a larger issue described by #1124.